### DISCUSSION TITLE
How to override the error serializer/sanitizer?

### DISCUSSION THREAD

**Discussion by adaboese** [remix-run/remix#8454] [Q&A]
Documentation mentions sanitization:

> If you need to log these errors or report then to a third-party service such as [BugSnag](https://www.bugsnag.com/) or [Sentry](https://sentry.io/), then you can do this through a [handleError](https://remix.run/docs/en/main/file-conventions/entry.server#handleerror) export in your [app/entry.server.js](https://remix.run/docs/en/main/file-conventions/entry.server). This method receives the un-sanitized versions of the error since it is also running on the server.

https://remix.run/docs/en/main/guides/errors#error-sanitization

Where is this logic implemented and how do I override it?

My use case is that I need to attach an additional property to the error.

Namely, when an error occurs, I want to generate Sentry error ID and display that to the user.

```ts
import { captureException } from '@sentry/node';

// ...

const sentryErrorId = captureException(error);
```

I would imagine that I can do this in the error sanitization stage, e.g.

```ts
const sanitizeError = (error) => {
  const sentryErrorId = captureException(error);

  const sanitizedError = originalSanitizeError(error);

  sanitizedError.sentryErrorId = sentryErrorId;

  return sanitizedError;
};
```

---

I just realized that if I do this, then I will get errors reported twice, because `@sentry/remix` already wraps error handler. I can always remove it, but just flagging that there is this issue.

**Comment by steuraa** [comment]
I'm wondering about this as well. 
In our project we're trying to show extra info about the occurred error (via error codes coming from the backend) in our error boundary and simultaneously log the errors in Sentry. 
It would be nice if there was an option to pass any extra data to the Errorboundary when throwing an Error, or to disable/amend the sanitizing.

**Reply by kiliman** [reply]
Yes, I had that issue as well. Unfortunately, the only thing that gets through to the error boundary is `error.message` (and stack in development) or `error.data` for thrown responses.

In my `remix-global-data` example I work around it by setting `.message` to `JSON.stringify(data)` in `handleError` and then parse it in my error boundary.

https://github.com/kiliman/remix-global-data#access-request-data-in-errorboundary

**Reply by steuraa** [reply]
Thanks @kiliman , tried it out and seems like a simple/workable solution :+1:

**Reply by thomasverleye** [reply]
@kiliman After testing this out, it seems like overwriting the `.message` aint possible after all.
We consistently are receiving the error `Error: Unexpected Server Error` while in the `handleError` we're overwriting the `error.message` 🤔 

Did something change in the way Remix sanitize the errors?

**Reply by gustavopch** [reply]
This is the relevant code:
https://github.com/remix-run/remix/blob/ede2012949a784722d8989124dff5a34ae90c008/packages/remix-server-runtime/errors.ts#L48-L55

**Comment by manzaloros** [comment]
[Started a conversation about this here](https://github.com/remix-run/react-router/discussions/13213)

**Comment by jdevcc** [comment]
The ability to customize error serialization would solve my [React Router v7 issue](https://github.com/remix-run/react-router/discussions/13314).
